### PR TITLE
Waypoint editing

### PR DESCRIPTION
**Pull request by exofunctor**
Changes:
- Added colliders to buttons in UI
- Deactivated button animators. We should instead temporarily change the material instead, when they are pressed, to avoid the bug where they get stuck.
- Unused scripts went to deprecated folder
- All scripts now use getter methods instead of the waypoints array to do waypoint operations

- Merged master
- Added colliders on sensor UI
- Fixed error in drone sim landing button
- Sensor manager null reference check
- Sensor manager delete old list
- Sensor manager converted list to a dictionary to keep track of both subscribed and unsubscribed topics 
- Update M210_Flight_test_manager automatically.

### CODE REVIEW

**Review by PeruDayani — COMMENTED** [review]
Looks good, I'll add some changes and documentation on my end.

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 1]
```diff
@@ -260,33 +288,33 @@ public int WaypointsCount()
         }
 
         /// <summary>
-        /// Return the number of deleted waypoints
+        /// The number of waypoints that were placed, but got deleted for the current drone.
```
Comment (erwang01) [inline Assets/Scripts/Drone Scripts/Drone.cs:291]: Does the script make a difference between current drone and non current drone?

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 2]
```diff
@@ -260,33 +288,33 @@ public int WaypointsCount()
...
                 return 0;
             }
-            return deletedWaypoints.Count;
+            return deletedWaypoints.Count; 
         }
 
+
         /// <summary>
-        /// Return true if waypoint list is empty and false if not
+        /// TODO: document
```
Comment (erwang01) [inline Assets/Scripts/Drone Scripts/Drone.cs]: todo

File: Assets/Scripts/Interaction Scripts/ControllerInput.cs [hunk 3]
```diff
@@ -273,6 +301,32 @@ public Transform LeftUITransform()
 		//  RIGHT HAND GETTER METHODS  //
 		/*******************************/
 		
+		/// <summary>
+		/// Whether the right hand is currently touching an object.
+		/// </summary>
+        /// <returns>True if the right hand is currently touching an object.</returns>
```
Comment (erwang01) [inline Assets/Scripts/Interaction Scripts/ControllerInput.cs:307]: missing a tab.

File: Assets/Scripts/Menu Scripts/SensorManager.cs [hunk 4]
```diff
@@ -13,9 +13,10 @@ public class SensorManager : MonoBehaviour {
     public Button rightButton;
     public Text sensorText;
     public GameObject togglePrefab;
+    public GameObject sensorMenu;
 
     private List<ROSSensorConnectionInterface> sensorList = new List<ROSSensorConnectionInterface>();
-    private List<string> subscriberList = new List<string>();
+    private Dictionary<string, bool> subscriberDict = new Dictionary<string, bool>();
```
Comment (erwang01) [inline Assets/Scripts/Menu Scripts/SensorManager.cs:19]: what is the purpose of the bool?

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs [hunk 5]
```diff
@@ -20,7 +20,7 @@ public class MeshSensor_ROSSensorConnection : MonoBehaviour, ROSTopicSubscriber,
     private ROSBridgeWebSocketConnection ros = null;
     public string client_id;
     private Thread rosMsgThread;
-    private List<string> sensorSubscriberTopics = new List<string>();
+    private Dictionary<string, bool> sensorSubscriberTopicsDict = new Dictionary<string, bool>();
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs:23]: Can we add the dict to the ROSTopicSubscriber?

**Review by erwang01 — APPROVED** [review]

File: Assets/Scripts/Drone Scripts/Drone.cs (hunk 1 above)
Comment (PeruDayani) [reply Assets/Scripts/Drone Scripts/Drone.cs:291]: yep, each drone has it's own drone.cs script

File: Assets/Scripts/Menu Scripts/SensorManager.cs (hunk 4 above)
Comment (PeruDayani) [reply Assets/Scripts/Menu Scripts/SensorManager.cs:19]: to store the status of the subscription. if it's subscribed or not currently.

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs (hunk 5 above)
Comment (PeruDayani) [reply Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs:23]: What do you mean?

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 6]
```diff
@@ -242,14 +247,37 @@ public void DeleteAllWaypoints()
...
         /// </summary>
+        /// <param name="waypointList"></param>
         /// <returns></returns>
+        private bool isEmptyWaypointList(List<Waypoint> waypointList)
+        {
+            if (waypointList == null)
+            {
+                return true;
+            }
+            return waypointList.Count == 0;
```
Comment (jasmine-bae) [inline Assets/Scripts/Drone Scripts/Drone.cs:267]: When would there be a case where the list is empty but it's count is more than 0?

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 7]
```diff
@@ -260,33 +288,33 @@ public int WaypointsCount()
...
                 return 0;
             }
-            return deletedWaypoints.Count;
+            return deletedWaypoints.Count; 
         }
 
+
         /// <summary>
-        /// Return true if waypoint list is empty and false if not
+        /// TReturn the waypoint at the requested index if valid
```
Comment (jasmine-bae) [inline Assets/Scripts/Drone Scripts/Drone.cs:305]: Small typo with "Return" comment

File: Assets/Scripts/Interaction Scripts/ControllerInput.cs [hunk 8]
```diff
@@ -65,6 +66,32 @@ void Awake()
...
+		/// <summary>
+		/// Whether the left hand is currently touching an object.
+		/// </summary>
+        /// <returns>True if the left hand is currently touching an object.</returns>
+		public bool LeftIsTouchingWaypoint()
+		{
+			GameObject touched = LeftController.GetComponent<VRTK_InteractTouch>().GetTouchedObject();
+			if (touched)
+			{
+				if (touched.tag == "waypoint")
```
Comment (jasmine-bae) [inline Assets/Scripts/Interaction Scripts/ControllerInput.cs:87]: Can't we combine these two if statements (i.e. touched && (touched.tag == "waypoint"))?

Comment (jasmine-bae) [reply Assets/Scripts/Interaction Scripts/ControllerInput.cs:87]: or just `return touched && (touched.tag == "waypoint")`

File: Assets/Scripts/Interaction Scripts/ControllerInput.cs [hunk 9]
```diff
@@ -81,7 +108,8 @@ public bool LeftIsGrabbing()
 		public bool LeftIsGrabbingWaypoint()
 		{
 			GameObject grabbed = LeftController.GetComponent<VRTK_InteractGrab>().GetGrabbedObject();
-			if (grabbed) {
+			if (grabbed)
+			{
```
Comment (jasmine-bae) [inline Assets/Scripts/Interaction Scripts/ControllerInput.cs:112]: Same with grabbed here , just `return grabbed && (grabbed.tag == "waypoint")`

File: Assets/Scripts/Interaction Scripts/ControllerInput.cs [hunk 10]
```diff
@@ -273,6 +301,32 @@ public Transform LeftUITransform()
...
+			if (touched)
+			{
+				if (touched.tag == "waypoint")
+				{
+					return true;
+				}	
+			}
+			return false;
+		}
+
```
Comment (jasmine-bae) [inline Assets/Scripts/Interaction Scripts/ControllerInput.cs:329]: Alternatively just keep the same kind of if statement/return/ternary statement throughout these very similar methods.

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/PointCloudSensor_ROSSensorConnection.cs [hunk 11]
```diff
@@ -35,7 +35,9 @@ public void InitilizeSensor(int uniqueID, string sensorIP, int sensorPort, List<
             pcVisualizer.SetParentTransform(this.transform);
             pcVisualizers.Add(subscriberTopic, pcVisualizer);
             Debug.Log("Point Cloud subscribing to: " + subscriberTopic);
-            ros.AddSubscriber("/voxblox_node/" + subscriber, this);
+
+            sensorSubscriberTopicsDict.Add(subscriberTopic, true);
+            ros.AddSubscriber(subscriberTopic, this);
```
Comment (jasmine-bae) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/PointCloudSensor_ROSSensorConnection.cs:40]: the subscriber topic won't work by itself, needs to have "/voxblox_node/" in front (simlarly with camera streaming sensors I'm working on)

**Review by jasmine-bae — COMMENTED** [review]
Seems good, just some issues with typos and stuff + one bug in Ros Connection Interfaces.

File: Assets/Scripts/Drone Scripts/Drone.cs (hunk 6 above)
Comment (PeruDayani) [reply Assets/Scripts/Drone Scripts/Drone.cs:267]: there might be a case initially where the list variable is null, therefore it's done this way.
@apollovision any other reason?

File: Assets/Scripts/Interaction Scripts/ControllerInput.cs (hunk 8 above)
Comment (PeruDayani) [reply Assets/Scripts/Interaction Scripts/ControllerInput.cs:87]: good point, will do

**Pull request merged**